### DISCUSSION TITLE
Using grep crate (as library)

### DISCUSSION THREAD

**Discussion by bq-wrongway** [BurntSushi/ripgrep#2509] [Q&A]
Hi, 
i am having an issue that i am not sure how to solve :
```
    let matcher = RegexMatcher::new_line_matcher(&pattern)?;
    let mut searcher = SearcherBuilder::new().line_number(true).build();
```

I use something like this to create a searcher , and i use iced.rs to provide gui for search, 

Then from UI, i select some directory or file, put in a word, and perform search which i can either save or present, even piping works, i can search just FOO or i can search FOO|BAR and it will find both words, but i dont see way to tell Searchbuilder (or matcher) to search ONLY rust files for example , -trust wont work.

I am not sure how would i be able to search only certain files, also going to grep create and checking grep_regex, grep_searcher i did not find way to add some extension filtering.
with cli is trivial, but with library is bit unclear.

Thank you

**Comment by BurntSushi** [comment]
Yeah the high level documentation for "ripgrep as a library" is basically non-existent.

The problem here is that the `grep` family of crates are just for searching something. They don't help you with gathering what to search. You can do whatever you want to gather what to search. `walkdir` is a pretty standard implementation for directory traversal and you can apply your own filter rules at that level. The `ignore` crate is much higher level, uses parallelism, respects your `.rgignore` and `.gitignore` files and you can specify custom globs for filtering.
